File: 784.letter-case-permutation.py
#
# @lc app=leetcode id=784 lang=python3
#
# [784] Letter Case Permutation
#

# @lc code=start


from collections import deque


class Solution:
    def letterCasePermutation(self, s: str) -> List[str]:

        # def backtrack(sub="", i=0):
        #     if len(sub)==n:
        #         res.append(sub)
        #     elif s[i].isdigit():
        #         backtrack(sub+s[i], i+1)
        #     else:
        #         backtrack(sub+s[i].swapcase(), i+1)
        #         backtrack(sub+s[i],i+1)
        # backtrack()

        res = []
        stack = deque([("", 0)])
        n = len(s)

        while stack:
            sub, i = stack.pop()
            if len(sub) == n:
                res.append(sub)
            elif s[i].isdigit():
                stack.append((sub + s[i], i + 1))
            else:
                stack.append((sub + s[i].swapcase(), i + 1))
                stack.append((sub + s[i], i + 1))

        return res


# @lc code=end

File: 784.letter-case-permutation.rs
/*
 * @lc app=leetcode id=784 lang=rust
 *
 * [784] Letter Case Permutation
 */

// @lc code=start
use std::collections::VecDeque;
impl Solution {
    pub fn letter_case_permutation(s: String) -> Vec<String> {
        let (mut res, mut substr) = (vec![], VecDeque::new());

        fn dfs(substr: &mut VecDeque<char>, s: &str, res: &mut Vec<String>) {
            let sublen = substr.len();
            if sublen == s.len() {
                res.push(substr.iter().collect());
            } else if let Some(c) = s.chars().nth(sublen) {
                if c.is_ascii_alphabetic() {
                    substr.push_back(c.to_ascii_lowercase());
                    dfs(substr, s, res);
                    substr.pop_back();

                    substr.push_back(c.to_ascii_uppercase());
                    dfs(substr, s, res);
                    substr.pop_back();
                } else {
                    substr.push_back(c);
                    dfs(substr, s, res);
                    substr.pop_back();
                }
            }
        }

        dfs(&mut substr, &s, &mut res);
        res
    }
}
// @lc code=end
